ArmPkg/Drivers/SmmuDxe/IoMmu.c

```diff
@@ -0,0 +1,523 @@
+/** @file IoMmu.c
+
+    This file contains functions for the IoMmu protocol.
+
+    Copyright (c) Microsoft Corporation. All rights reserved.
+    SPDX-License-Identifier: BSD-2-Clause-Patent
+
+**/
+
+#include <Uefi.h>
+#include <Library/ArmLib.h>
+#include <Library/BaseMemoryLib.h>
+#include <Library/DebugLib.h>
+#include <Library/IoLib.h>
+#include <Library/MemoryAllocationLib.h>
+#include <Library/TimerLib.h>
+#include <Library/UefiBootServicesTableLib.h>
+#include <Library/UefiDriverEntryPoint.h>
+#include <Protocol/IoMmu.h>
+#include "IoMmu.h"
+#include "SmmuV3Registers.h"
+
+#define PAGE_TABLE_DEPTH  4                             // Number of levels in the page table

```

Comment:
Typically, generic macros like these would go in a header file.

Reply:
moved to IoMmu.h